Code/Framework/AzCore/AzCore/Jobs/JobExecutor.cpp

```diff
@@ -0,0 +1,349 @@
+                AZStd::thread_desc desc = {};
+                desc.m_name = threadName.c_str();
+                if (affinitize)
+                {
+                    desc.m_cpuId = 1 << id;
+                }
+                m_active.store(true, AZStd::memory_order_release);
+
+                m_thread = AZStd::thread{ [this, &initSemaphore]
+                                          {
+                                              initSemaphore.release();
+                                              Run();
+                                          },
+                                          &desc };
+            }
+
+            void Join()
+            {
+                m_active.store(false, AZStd::memory_order_release);
+                m_semaphore.release();
+                m_thread.join();
+            }
+
+            void Enqueue(TypeErasedJob* job)
+            {
+                m_queue.Enqueue(job);
+
+                if (!m_busy.exchange(true))
+                {
+                    // The worker was idle prior to enqueueing the job, release the semaphore
+                    m_semaphore.release();
+                }
+            }
+
+        private:
+            void Run()
+            {
+                while (m_active)
+                {
+                    m_busy = false;
+                    m_semaphore.acquire();
+
+                    if (!m_active)
+                    {
+                        return;
+                    }
+
+                    m_busy = true;
+
+                    TypeErasedJob* job = m_queue.TryDequeue();
+                    while (job)
+                    {
+                        job->Invoke();
+                        // Decrement counts for all job successors
+                        for (size_t j = 0; j != job->m_outboundLinkCount; ++j)
+                        {
+                            TypeErasedJob* successor = job->m_graph->m_successors[job->m_successorOffset + j];
+                            if (--successor->m_dependencyCount == 0)
+                            {
+                                m_executor->Submit(*successor);
+                            }
+                        }
+
+                        job->m_graph->Release();
+                        --m_executor->m_remaining;
+
+                        job = m_queue.TryDequeue();
+                    }
+                }
+            }
+
+            AZStd::thread m_thread;
+            AZStd::atomic<bool> m_active;
+            AZStd::atomic<bool> m_busy;
+            AZStd::binary_semaphore m_semaphore;
+
+            ::AZ::JobExecutor* m_executor;
+            JobQueue m_queue;
+        };
+    } // namespace Internal
+
+    JobExecutor& JobExecutor::Instance()
+    {
+        // TODO: Create the default executor as part of a component (as in JobManagerComponent)
+        static JobExecutor executor;
+        return executor;
+    }
```

Comment:
This could use an `AZ::Environment<JobExecutor>` instead.

Otherwise as how it is architected now, this will not work across dll boundaries.

Reply:
Agreed, I intend on using `AZ::Environment` and mimicking the existing pattern of initializing it during the system component bootup sequence.

Reply:
In that case you might want to use `AZ::Interface` instead. That already uses the `AZ::Environment` behind the scenes.